### PR TITLE
lutris: init at 0.4.18

### PR DESCRIPTION
**Pull request by ghost**
###### Motivation for this change
Steam-like program for keeping games in order and easily accessible.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Chiiruno** [comment]
`nix-shell -p nox --run "nox-review wip"` errored, however the error was an osquery test, Lutris itself seems to work fine, aside from this issue when using online functions I'm not sure how to resolve:
```
Traceback (most recent call last):
  File "/nix/store/6864ay4b9skqskzi4sszbyx8pycwy4pk-lutris-v0.4.18/lib/python3.6/site-packages/lutris/gui/lutriswindow.py", line 169, in <lambda>
    lambda *x: open_uri('https://lutris.net/user/register/')
  File "/nix/store/6864ay4b9skqskzi4sszbyx8pycwy4pk-lutris-v0.4.18/lib/python3.6/site-packages/lutris/util/system.py", line 387, in open_uri
    Gtk.show_uri(None, uri, Gdk.CURRENT_TIME)
GLib.Error: g-io-error-quark: Operation not supported (15)
```

**Review by lukateras — APPROVED** [review]
LGTM!

**Comment by yegortimoshenko** [comment]
This should work for the most part, although native Steam client won't be accessible.

**Comment by yegortimoshenko** [comment]
Oh, have you tested if external emulators (say, SNES9x) installed either in user or system profile work?

**Comment by bkchr** [comment]
Any plan to maybe integrate a patch, that calls `patchelf` on dowloaded wine installations?

**Comment by yegortimoshenko** [comment]
There's no need, because it runs in a FHS environment that tolerates ELF headers of downloaded Wine installations

**Comment by bkchr** [comment]
Ahh! Nice! I did not saw that, then it is really nice :)

**Comment by Chiiruno** [comment]
The native steam client should be possible by adding steam-runtime-native to the FHS environment IIRC.
I can't tell one way or another if the external emulators work, I do know they download and I can log in, but adding any games to the menu is either impossible or impossible to show, there are no error messages, just no action.

Sorry, I guess I pushed a somewhat broken commit in my excitement.

**Comment by Chiiruno** [comment]
Again, a lot of the trouble probably stems from this:
```
Traceback (most recent call last):
  File "/nix/store/6864ay4b9skqskzi4sszbyx8pycwy4pk-lutris-v0.4.18/lib/python3.6/site-packages/lutris/gui/lutriswindow.py", line 169, in <lambda>
    lambda *x: open_uri('https://lutris.net/user/register/')
  File "/nix/store/6864ay4b9skqskzi4sszbyx8pycwy4pk-lutris-v0.4.18/lib/python3.6/site-packages/lutris/util/system.py", line 387, in open_uri
    Gtk.show_uri(None, uri, Gdk.CURRENT_TIME)
GLib.Error: g-io-error-quark: Operation not supported (15)
```

**Comment by Chiiruno** [comment]
Okay, will probably have to pull dependencies like X.

**Comment by tenten8401** [comment]
It seems regardless of distro, Lutris needs a restart for my games to properly show up, at least when syncing from the lutris website. I don't think that's related to anything you've done.

**Comment by tenten8401** [comment]
Getting a fun error over here when attempting to install Battlefield 1 using the Wine runner.
image

**Comment by Chiiruno** [comment]
This is a big project with a lot of "unknown" dependencies.
Please report all bugs and I'll attempt to fix them.
Will continue until I get everything (at least on the surface) working.

**Comment by Chiiruno** [comment]
@tenten8401 Could you give more detail about that error please?
Does something need to be added into the target packages?

**Comment by tenten8401** [comment]
Looks like for some reason it couldn't find the utility to make an sha1sum
and proceeded to try and wipe any install files it made in /tmp

**Comment by Chiiruno** [comment]
What package is sha1sum part of? Isn't it bin or coreutils?

**Comment by tenten8401** [comment]
I think it's coreutils, could be wrong

**Comment by Chiiruno** [comment]
Could you test it out? I don't have BF1.
Just add coreutils to `targetPkgs = pkgs: with pkgs; with xorg; [`

**Comment by tenten8401** [comment]
You don't need BF1 to test it I don't think, I haven't even gotten to the
part where I enter my origin login. I'll give it shot shortly.

**Comment by Chiiruno** [comment]
Okay, that commit should at least make this somewhat usable.
As far as I can tell, most emulators work, I can't get WINE to work no matter what I do, same with this error on Citra: `This application failed to start because it could not find or load the Qt platform plugin "xcb"` - It's a QT issue or some sort. Mednafen is weird with sound, not the fault of this package.

**Comment by yegortimoshenko** [comment]
@Chiiruno `export QT_QPA_PLATFORM_PLUGIN_PATH=$(nix-build '<nixpkgs>' -A qt5.qtbase)/lib/qt-5.10/plugins/platforms` will likely help (could be part of `buildFHSUserEnv` `runScript`).

**Comment by tenten8401** [comment]
Seems like sha1sum is actually a winetricks dependency rather than a lutris
dependency. Lutris ships it's own build of winetricks with the client.

Also, just a PSA: I'd test out the lutris package in a VM, as it seems like
winetricks doesn't like it when it's missing dependencies and seems to have
wiped my /home folder, last screenshot I uploaded seems to be related, as
lutris doesn't use RM through bash.

Currently running testdisk on my drive to see if I can recover it. Looks
like most of the files are recoverable however since nothing else got
written before I shut my computer off.

**Comment by Chiiruno** [comment]
I'm not sure there's a lot more I can do to improve this package, any idea?

**Comment by yegortimoshenko** [comment]
I would like to investigate wipe that @tenten8401 had before committing package to the tree.

**Comment by Chiiruno** [comment]
Yeah, I'm worried about that too.
Rather, I didn't know that was even possible on Nix.

**Comment by tenten8401** [comment]
Got my system back up and running. Didn't loose basically anything.

I'll see if I can make a VM and see if I can reproduce that again.

**Comment by tenten8401** [comment]
Seems like as of this patch there's an undefined variable "pulseaudio" in the default.nix, probably related to the extra dependencies you added.

**Comment by tenten8401** [comment]
Pretty sure this is how winetricks ended up wiping my home folder last time.
img
image
Seems to want to wipe /tmp without checking what it actually put in there.

**Comment by yegortimoshenko** [comment]
Interesting! This means we shouldn't bind-mount `/tmp`.

**Comment by Chiiruno** [comment]
I'm still going through each thing there is to install and getting as much working as possible, but I don't think I can get everything 100%, at least not without some help.

**Comment by Chiiruno** [comment]
@tenten8401 Could you try again with this new commit to see if you can still reproduce the problem?

**Comment by Chiiruno** [comment]
Is there a way to add the current system's libraries and executables to the FHS environment?
That would massively simplify this.

**Comment by Chiiruno** [comment]
That should get most things working, anything else, such as the qt xcb issues, I can't seem to debug.
I'll keep trying, of course.

**Comment by tenten8401** [comment]
One thing I notice is the lack of the `gtksourceview` package. Perhaps you meant `gnome(3/2).gtksourceview`?

I'm also getting this strange error:
image

**Comment by Chiiruno** [comment]
@tenten8401 Tell me if that fixes your `gtksourceview` issue please.
I haven't the faintest idea what to do about the error in the image, sorry.
I don't think it's a problem with this package though from some cursory research, maybe try a 64 bit prefix and install .NET 4.5?

**Comment by tenten8401** [comment]
Didn't exactly fix the issue as you forgot to remove gtksourceview from where it was before.

I think the "Taskset/Cpuset not available on your platform!" issue may be a symptom of running it in a virtual machine. I'll make a full disk backup and try it on real hardware.

**Comment by Chiiruno** [comment]
@tenten8401 Okay, try again please.

**Comment by tenten8401** [comment]
Still getting the taskset/cpuset error. Not a clue what could be causing it.
image

**Comment by Chiiruno** [comment]
I can't say I know, but it's likely related to the fact that we need to incorporate the system environment with the FHS environment somehow.
I'm at a bit of a loss, so maybe someone will come help.

**Comment by Chiiruno** [comment]
@pstn I'm not sure why you deleted your comment, but I tested both the unaltered package and a rebased-to-upstream version and they both build and run fine.
I'm not entirely sure why your build failed.
Since you deleted your comment, I assume you got it working?

**Comment by pstn** [comment]
Yeah, was completely my fault and unrelated to the package. Didn't want to make any fuzz about it.

**Comment by tenten8401** [comment]
After messing around with wine some, I realized what may be the issue with the taskset/cpuset invalid problem.

Turns out, the default Wine package in NixOS requires an override for 64-bit application support. This may be the actual problem with the game I was trying to install, which would explain the invalid platform issues.
https://nixos.wiki/wiki/Wine

**Comment by Chiiruno** [comment]
So we need to use 64bit WINE, or maybe WINE-WOW?

**Comment by Chiiruno** [comment]
Please tell me if this works for you.

**Comment by tenten8401** [comment]
I'll give it a go.

**Comment by tenten8401** [comment]
Nope, still have the taskset/cpuset not available on your platform error sadly.

**Comment by tenten8401** [comment]
So I enabled debug logging and it looks like it's failing to create the prefix in my home folder for some reason... https://hastebin.com/odamukited.sql

**Comment by tenten8401** [comment]
Alright, so the taskset/cpuset issue is fixed by installing utillinux, but apparently that was only a warning and not an actual issue.

**Comment by tenten8401** [comment]
The main error I'm seeing is this:
`shell-init: error retrieving current directory: getcwd: cannot access parent directories: No such file or directory`

I don't know if it's related to why the log window is now empty but if I had to guess, that'd be the main problem.

**Comment by tenten8401** [comment]
Just going to recap my findings when letting it go for a while:
- load_builtin_dll failed to load .so lib for builtin L"winex11.drv": libXext.so.6: cannot open shared object file: No such file or directory
- Wine cannot find the FreeType font library.  To enable Wine to use TrueType fonts please install a version of FreeType greater than or equal to 2.0.5.
- load_builtin_dll failed to load .so lib for builtin L"WLDAP32.dll": libldap_r-2.4.so.2: cannot open shared object file: No such file or directory
- load_builtin_dll failed to load .so lib for builtin L"WININET.dll": libz.so.1: cannot open shared object file: No such file or directory

**Comment by tenten8401** [comment]
I'm not trying to install any steam games, but that might be good to do. Perhaps we also need to add the Wine libraries to the LD_LIBRARY_PATH since Lutris installs it's own version of Wine, which isn't patched?

**Comment by tenten8401** [comment]
Yeah, I think it's related to the LD_LIBRARY_PATH. If I tell Lutris to use the version of Wine that's included in the package instead of the one it's downloaded, it seems to work as expected.

**Comment by Chiiruno** [comment]
What should we do, then? I'm not very savvy on library paths.
Just add `${pkgs.whatever}/lib` to the LD_LIBRARY_PATH?

**Comment by tenten8401** [comment]
I have no clue, but that sounds like it might work.

**Review by matthewbauer — APPROVED** [review]
All of this looks good! My only complaint would be to maybe squash a few of the commits together (so each new package is one commit). Otherwise please make sure everything is working properly and this should be mergeable.

**Comment by Chiiruno** [comment]
Well, that's the thing. Not everything is working properly.
Most of lutris seems to work, but it's optional downloaded programs have issues, and we need help figuring out how to give it the FHS environment it needs.
Is there a way to tack on the system environment to the FHS environment?

**Comment by tenten8401** [comment]
I don't think it's related to the system environment at all. I think it's the fact that Lutris ships their own non-patched versions of Wine, which can't find the required libraries to function. If I use the installed version of Wine that comes with the package instead of the Lutris built ones, my stuff works.

**Comment by Chiiruno** [comment]
WINE isn't the only issue. Try launching something like Citra.

**Comment by tenten8401** [comment]
also, dumb question, but is makeWrapper even getting called? I see you have the args but I don't know if it's getting called or not, as I don't see anywhere the LD_LIBRARY_PATH is being set.

**Comment by tenten8401** [comment]
Alright, I'm at the point where my lack of knowledge of the Nix package management system is what's stopping me from working on this further.

I'd be willing to send $10 PayPal to anyone who can successfully get this package fully working as intended, since I think this point we're basically stuck on Lutris versions of Wine not being able to find the libraries correctly.

**Comment by Chiiruno** [comment]
#44047 may be a possible fix for the QT issues.

**Comment by Infinisil** [comment]
While trying to install Portal 1:

```
WARNING  2018-10-13 23:05:43,893 [lutriswindow]:portal is not available
INFO     2018-10-13 23:05:57,276 [steam]:Data path for SteamApp 400 not found.
** (process:25779): WARNING **: 23:05:57.289: chrootenv doesn't stack!
** Message: 23:05:57.290: Requires Linux version >= 3.19 built with CONFIG_USER_NS
** (process:25781): ERROR **: 23:05:57.291: main: unshare: Operation not permitted
```

Which sounds pretty bad.. Seems like we don't have a chance for this to ever work reliably, because we need to run everything in a chroot to begin with.

**Comment by Chiiruno** [comment]
So there's nothing we can do?
What about how the Steam package does things?
Can we incorporate anything from other FHS packages?

**Comment by colemickens** [comment]
Has anyone tried it with a kernel with `CONFIG_USER_NS` enabled? Some other distros do enable it.

**Comment by Infinisil** [comment]
@colemickens Hmm I thought it was enabled, even by default, but now I'm not so sure anymore. I originally interpreted the `Requires Linux version >= 3.19 built with CONFIG_USER_NS` error as just being a false assumption based on the first error by the program.

But now I'm also confused even more, because in [the man page on user namespaces](http://man7.org/linux/man-pages/man7/user_namespaces.7.html) they clearly mention that they can be nested up to 32 levels deep. I guess `chrootenv doesn't stack!` doesn't have much to do with user namespaces maybe? Not sure what `stack` means in this context if not nesting.

**Comment by yegortimoshenko** [comment]
@colemickens It's on for NixOS users unless one uses hardened kernel:

```
$ zcat /proc/config.gz | grep CONFIG_USER_NS
CONFIG_USER_NS=y
```

@Infinisil Good find!. Now we should try to understand why it fails to unshare the second time. Source code for reference: https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/build-fhs-userenv/chrootenv/chrootenv.c

**Comment by JohnAZoidberg** [comment]
It doesn't work for me anymore.
Most (all?) curls to `https://abbradar.net/steamrt/pool/main/...` and `http://repo.steampowered.com/steamrt/pool/main/...` result in 404.

**Comment by davidak** [comment]
@JohnAZoidberg that's related to https://github.com/NixOS/nixpkgs/issues/49048

File: pkgs/applications/misc/lutris/default.nix [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+    propagatedBuildInputs = deps;
+    python2Path = python2Deps;
+    fullPath = stdenv.lib.makeLibraryPath deps;
+    preConfigure = "export HOME=$PWD";
+
+    makeWrapperArgs = [
+      "--prefix LD_LIBRARY_PATH : ${fullPath}:$out/lib"
+      "--set GI_TYPELIB_PATH $GI_TYPELIB_PATH"
+      "--prefix XDG_DATA_DIRS : $out/share"
+      "--suffix XDG_DATA_DIRS : $XDG_ICON_DIRS:$GSETTINGS_SCHEMAS_PATH"
```
Comment (Mic92) [inline pkgs/applications/misc/lutris/default.nix]: GSETTINGS_SCHEMAS_PATH will be also set automatically by `wrapGAppsHook` when `glib gnome3.gsettings_desktop_schemas` are in `buildInputs`: https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/networking/instant-messengers/rambox/default.nix#L33

**Comment by nyanloutre** [comment]
Would a similar approach work for GameHub, or did anyone tried to package it ?

https://github.com/tkashkin/gamehub

**Comment by Laaas** [comment]
What are the issues with this PR?

**Comment by tycho01** [comment]
we may want to update the PR at least -- using the present one I just ran into https://github.com/lutris/lutris/issues/967.

**Comment by tycho01** [comment]
trying to [update](https://github.com/tycho01/nixpkgs/compare/7c6f263ba794bfd363789f812192685f4e037876...tycho01:lutris) it, but still figuring out how to expose `requests` to it:
```
  Could not find a version that satisfies the requirement requests (from lutris==0.5.1) (from versions: )
No matching distribution found for requests (from lutris==0.5.1)
```

**Comment by teto** [comment]
@tycho01 that might come from a version mismatch, you can fetch a requests from master or patch the setup.py with the hope it works.
The latest lutris allows to access the epic store which has made a few interesting games free in the upcoming weeks (thinking about Assassin's Creed Unity where you can jump from Notre Dame):
https://www.gamingonlinux.com/articles/you-can-now-easily-run-the-epic-store-on-linux-with-lutris-epic-suggests-applying-for-a-grant.13964

**Comment by TheSirC** [comment]
It seams that the only conflicting files on this PR is on the top-level packages file. Any updates on the PR other than that ?

**Comment by abbradar** [comment]
I've rebased and updated that PR at https://github.com/NixOS/nixpkgs/pull/61742 and fixed Steam too. I noticed that some people offered bounty for this but I'm here just for the fun ;)

**Comment by Chiiruno** [comment]
Closed in favor of #61742

**Pull request closed without merging**